sql/src/main/java/io/crate/operation/collect/collectors/RemoteCollectBatchIterator.java

```diff
@@ -0,0 +1,347 @@
+                throw new IllegalStateException("BatchIterator is loading");
+            }
+            return false;
+        }
+        return source.moveNext();
+    }
+
+
+    @Override
+    public void close() {
+        if (source != null) {
+            source.close();
+        }
+    }
+
+    @Override
+    public CompletionStage<?> loadNextBatch() {
+        if (sourceFuture == null) {
+            sourceFuture = new CompletableFuture<>();
+            remoteCollector.doCollect();
+            return sourceFuture;
+        } else {
+            return CompletableFutures.failedFuture(new IllegalStateException("BatchIterator is already loading"));
+        }
+    }
+
+    @Override
+    public boolean allLoaded() {
+        return sourceFuture != null;
+    }
+
+
+    static class RemoteCollector implements Killable {
+
+        private final static ESLogger LOGGER = Loggers.getLogger(RemoteCollector.class);
+        private final static int RECEIVER_PHASE_ID = 1;
+
+        private final UUID jobId;
+        private final String localNode;
+        private final String remoteNode;
+        private final TransportJobAction transportJobAction;
+        private final TransportKillJobsNodeAction transportKillJobsNodeAction;
+        private final JobContextService jobContextService;
+        private final RamAccountingContext ramAccountingContext;
+        private final BatchConsumer batchConsumer;
+        private final RoutedCollectPhase collectPhase;
+
+        private final Object killLock = new Object();
+        private JobExecutionContext context = null;
+        private Throwable killed = null;
+
+        RemoteCollector(UUID jobId,
+                        String localNode,
+                        String remoteNode,
+                        TransportJobAction transportJobAction,
+                        TransportKillJobsNodeAction transportKillJobsNodeAction,
+                        JobContextService jobContextService,
+                        RamAccountingContext ramAccountingContext,
+                        BatchConsumer batchConsumer,
+                        RoutedCollectPhase collectPhase) {
+            this.jobId = jobId;
+            this.localNode = localNode;
+            this.remoteNode = remoteNode;
+
+            this.transportJobAction = transportJobAction;
+            this.transportKillJobsNodeAction = transportKillJobsNodeAction;
+            this.jobContextService = jobContextService;
+            this.ramAccountingContext = ramAccountingContext;
+            this.batchConsumer = batchConsumer;
+            this.collectPhase = collectPhase;
+        }
+
+        public void doCollect() {
+            if (!createLocalContext()) return;
+            createRemoteContext();
+        }
+
+        @VisibleForTesting
+        boolean createLocalContext() {
+            JobExecutionContext.Builder builder = createPageDownstreamContext();
+            try {
+                synchronized (killLock) {
```

Comment:
Shall we replace synchronized with juc constructs ? ie. reentrantlock (synchronized sends the thread into a blocked state and can't really do much with it after that)